### PR TITLE
VPC block CSI driver crash due to null IksSession if IAM token exchange request fails when Storage-api-ms is unavailable

### PR DESCRIPTION
**Pull request by sameshai**

### CODE REVIEW

**Comment by sameshai** [comment]
Negative test case --

{“level”:“info”,“ts”:“2021-08-10T16:15:51.001+0530",“caller”:“utils/init_provider.go:91",“msg”:“Calling provider/utils/init_provider.go GenerateContextCredentials”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9"}
{“level”:“info”,“ts”:“2021-08-10T16:15:51.001+0530",“caller”:“utils/init_provider.go:106",“msg”:“Generating generateContextCredentials for “,”RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”,“Provider ID”:“iks-vpc-classic”}
{“level”:“info”,“ts”:“2021-08-10T16:15:51.001+0530”,“caller”:“provider/provider.go:68”,“msg”:“Entering IksVpcBlockProvider.OpenSession”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”}
{“level”:“info”,“ts”:“2021-08-10T16:15:51.001+0530”,“caller”:“provider/provider.go:73”,“msg”:“Opening VPC block session”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”}
{“level”:“info”,“ts”:“2021-08-10T16:15:51.001+0530”,“caller”:“provider/provider.go:75”,“msg”:“Its IKS dual session. Getttng IAM token for VPC block session”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”}
{“level”:“info”,“ts”:“2021-08-10T16:15:51.001+0530”,“caller”:“iam/token_exchange_iks.go:132”,“msg”:“In tokenExchangeIKSRequest’s sendTokenExchangeRequest()“,”RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”}
{“level”:“info”,“ts”:“2021-08-10T16:15:51.001+0530”,“caller”:“iam/token_exchange_iks.go:151”,“msg”:“Sending IAM token exchange request to container api server”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”}
{“level”:“info”,“ts”:“2021-08-10T16:15:53.467+0530”,“caller”:“provider/provider.go:177”,“msg”:“Entering OpenSession”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”}
{“level”:“info”,“ts”:“2021-08-10T16:15:53.467+0530”,“caller”:“provider/provider.go:200”,“msg”:“”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”,“apiConfig.ContextID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”}
{“level”:“info”,“ts”:“2021-08-10T16:15:53.467+0530”,“caller”:“metrics/metrics.go:73”,“msg”:“Time to complete”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9”,“OpenSession”:0.000066305}
{“level”:“info”,“ts”:“2021-08-10T16:15:53.467+0530",“caller”:“provider/provider.go:88",“msg”:“Opening IKS block session”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9"}
{“level”:“info”,“ts”:“2021-08-10T16:15:53.467+0530",“caller”:“provider/provider.go:99",“msg”:“Its ISK dual session. Getttng IAM token for IKS block session”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9"}
{“level”:“warn”,“ts”:“2021-08-10T16:15:53.467+0530",“caller”:“provider/provider.go:102",“msg”:“Error occurred while generating IAM token for IKS. But continue with VPC session alone. \n Volume Mount operation will fail but volume provisioning will work”,“RequestID”:“1b913077-be44-4617-a9b8-888bd6ab89b9"}
Enter the volume id: 11
Enter the instance id: 11
Enter the cluster id: 11
{"level":"warn","ts":"2021-08-10T16:16:01.977+0530","caller":"provider/attach_volume.go:49","msg":"Provider Session is not valid.","RequestID":"1b913077-be44-4617-a9b8-888bd6ab89b9"}
{"level":"info","ts":"2021-08-10T16:16:01.978+0530","caller":"metrics/metrics.go:73","msg":"Time to complete","RequestID":"1b913077-be44-4617-a9b8-888bd6ab89b9","AttachVolume":0.000171531}
{"level":"error","ts":"2021-08-10T16:16:01.978+0530","caller":"samples/attach_detach.go:51","msg":"Failed to attach the volume: ","RequestID":"1b913077-be44-4617-a9b8-888bd6ab89b9","error":"{Code:InvalidServiceSession, Type:RetrivalFailed, Description:The Service Session was not found due to error while generating IAM token for Sesssion Service., BackendError:IAM token exchange request failed,\"RequestID\":\"d5f38c32-cb9f-4e3b-8cd8-ff3b86335e62\",\"ResponseError\":\"json: unsupported type: func() (io.ReadCloser, error)\",\"error\":\"Error response from server. Status code: 503; message: upstream connect error or disconnect/reset before headers. reset reason: local reset, RC:500}"}

**Comment by sameshai** [comment]
Ideally we should return nil session for IKS if there is issue but today we can't do this as we have IKS and VPC session in same object. In future we need to separate out these for better error handling

**Review by meghnasiingh — APPROVED** [review]
LGTM

File: block/provider/attach_volume_test.go [hunk 1]
```diff
@@ -76,6 +179,80 @@ func TestAttachVolume(t *testing.T) {
 			assert.Nil(t, err)
 
 			volumeAttachService = &volumeAttachServiceFakes.VolumeAttachService{}
+			vpcs.APIClientVolAttachMgr = volumeAttachService
+			assert.NotNil(t, volumeAttachService)
+			uc.VolumeAttachServiceReturns(volumeAttachService)
+
+			if testcase.expectedErr != "" {
+				volumeAttachService.AttachVolumeReturns(testcase.baseVolumeAttachmentResponse, errors.New(testcase.expectedReasonCode))
+				volumeAttachService.ListVolumeAttachmentsReturns(testcase.baseVolumeAttachmentsListResponse, errors.New(testcase.expectedReasonCode))
```
Comment (arahamad) [inline block/provider/attach_volume_test.go:188]: can we optimize this if condition

File: block/provider/detach_volume_test.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
+/**
+ * Copyright 2020 IBM Corp.
```
Comment (arahamad) [inline block/provider/detach_volume_test.go:2]: change year to 2021

File: common/messages/messages_en.go [hunk 3]
```diff
@@ -219,6 +219,13 @@ var messagesEn = map[string]util.Message{
 		RC:          400,
 		Action:      "Please verify that the start volume ID is correct and whether you have access to the volume ID.",
 	},
+	"InvalidServiceSession": {
+		Code:        "InvalidServiceSession",
+		Description: "The Service Session was not found due to error while generating IAM token for Sesssion Service.",
```
Comment (arahamad) [inline common/messages/messages_en.go:224]: can it be `The Service Session was not found due to error while generating IAM token`

**Review by arahamad — CHANGES_REQUESTED** [review]
can you check these improvements

File: block/provider/attach_volume_test.go [hunk 4]
```diff
@@ -59,11 +54,114 @@ func TestAttachVolume(t *testing.T) {
...
+				}},
+			},
+
+			verify: func(t *testing.T, volumeAttachmentResponse *provider.VolumeAttachmentResponse, err error) {
+				assert.NotNil(t, volumeAttachmentResponse)
+				assert.Nil(t, err)
+			},
+		},
+		{
+			testCaseName: "Volume Attachment does not exist for the Volume ID- List Vol Attachement Fails",
```
Comment (sameshai) [inline block/provider/attach_volume_test.go:112]: Verify this if really required

File: block/provider/attach_volume_test.go [hunk 5]
```diff
@@ -99,3 +204,19 @@ func TestAttachVolume(t *testing.T) {
...
+	assert.NotNil(t, vpcs)
+	assert.NotNil(t, uc)
+	assert.NotNil(t, sc)
+	assert.Nil(t, err)
+	volumeAttachRequest := provider.VolumeAttachmentRequest{
+		VolumeID: "vol-1",
+	}
+
+	_, err = vpcs.AttachVolume(volumeAttachRequest)
+	assert.NotNil(t, err)
```
Comment (sameshai) [inline block/provider/attach_volume_test.go:221]: Error code validation

File: block/provider/detach_volume_test.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+				ID:         "16f293bf-test-4bff-816f-e199c0c65db5",
+				Href:       "",
+				Name:       "test volume name",
+				Status:     "stable",
+				Type:       "",
+				InstanceID: new(string),
+				ClusterID:  new(string),
+				Device:     &models.Device{},
+				Volume:     &models.Volume{ID: "volume-id1"},
+			},
```
Comment (sameshai) [inline block/provider/detach_volume_test.go:96]: Check this

File: block/provider/attach_volume_test.go [hunk 7]
```diff
@@ -59,11 +54,94 @@ func TestAttachVolume(t *testing.T) {
...
+				Status: "stable",
+				Volume: &models.Volume{ID: "volume-id1"},
+			},
+
+			baseVolumeAttachmentsListResponse: &models.VolumeAttachmentList{
+				VolumeAttachments: []models.VolumeAttachment{{
+					ID:     "16f293bf-test-4bff-816f-e199c0c65db5",
+					Name:   "test volume name",
+					Status: "stable",
+					Volume: &models.Volume{ID: "volume-id123"},
```
Comment (arahamad) [inline block/provider/attach_volume_test.go:137]: looks volume name need to be updated here

**Review by arahamad — APPROVED** [review]
looks to be some more improvement required in unit tests, will take that later

**Pull request merged**